# Numpy
- Numpy is an python library We call it a python library because it supports python syntax.
- It stands for numerical python.
- Numpy is used to perform advance numerical computations on large data sets.
- By using, Numpy we can perform complex operations with less line of code.
- Numpy provides very useful Data-structure known as Array.

### How to import Numpy ?

In [3]:
import numpy as np

In [3]:
arr = np.array([1,2,3,4,5])
print(arr , type(arr))

[1 2 3 4 5] <class 'numpy.ndarray'>


In [4]:
lavish = np.array(["lavish","gangwani",23,2.0] , dtype = 'object')
print(lavish , type(lavish))

['lavish' 'gangwani' 23 2.0] <class 'numpy.ndarray'>


In [5]:
lavish

array(['lavish', 'gangwani', 23, 2.0], dtype=object)

In [6]:
a= np.array([1,2,3,"45"])
a

array(['1', '2', '3', '45'], dtype='<U11')

In [7]:
a=np.array([1,2,3,"45"], dtype='object')
a

array([1, 2, 3, '45'], dtype=object)

### To Check the Version of your Numpy Library:

In [8]:
np.__version__

'1.24.3'

## ARRAY :
- It is an collection of Homogenous data.
   -  Example : np.array([1,2,3,4,5])
- As you can see all are in integer datatype.

In [9]:
arr = np.array([1,2,3,4,5,6,7,8])
print(arr , type(arr))

[1 2 3 4 5 6 7 8] <class 'numpy.ndarray'>


### List va Array:
- With array we can perform some operations which we can't perform with lists.
- Array are way faster than list.
- Which list if we have to print a range of make changes in the range then we have to write whole line of code.
- But with array we can simply do it.

In [10]:
lst=[i for i in range(1,12)]
print(lst)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [11]:
print(lst*5)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [12]:
# Multiply all the elements by 5.
lst=[i*5 for i in range(1,11)]
print(lst)

[5, 10, 15, 20, 25, 30, 35, 40, 45, 50]


In [13]:
arr=np.array(lst)
arr

array([ 5, 10, 15, 20, 25, 30, 35, 40, 45, 50])

In [14]:
arr*5


array([ 25,  50,  75, 100, 125, 150, 175, 200, 225, 250])

In [15]:
arr*7

array([ 35,  70, 105, 140, 175, 210, 245, 280, 315, 350])

#### Why Arays are faster than list:
- On list we can perform some operations.
- list make proccess slow.
- Array uses vectorisation method by using array broadcasting which make array was faster.
- List contains heterogenous data elements.
- Array contains of homogenous data elements.
- As Lists have Heterogenous data elements it has to store each element seprately by which
- As Lists have Heterogenous data elements it has to store each element seprately by which it makes proccess slow.
- As with Array have homogenous data elements so it cannot have to store elements seprately.

### int < float < complex < string < object

In [16]:
arr=np.array([1,2,3,4,5,6]) # intger
arr

array([1, 2, 3, 4, 5, 6])

In [17]:
arr=np.array([1,2.0,3,4,5]) #float
arr

array([1., 2., 3., 4., 5.])

In [18]:
arr=np.array([1,2,3+2j,4,5,6]) # complex
arr

array([1.+0.j, 2.+0.j, 3.+2.j, 4.+0.j, 5.+0.j, 6.+0.j])

In [19]:
arr = np.array([1,2,3,4,'5',6]) # string
arr

array(['1', '2', '3', '4', '5', '6'], dtype='<U11')

In [20]:
arr= np.array([1,2.0,3+4j,'4',5],dtype='object') # object
arr

array([1, 2.0, (3+4j), '4', 5], dtype=object)

In [21]:
arr.dtype

dtype('O')

### Create Different kinds of array :
###### np.arange()
   - It creates an array of numbers.
- Syntax - np.arange(start , stop , step)
    - start - inclusive (0 By default)
    - stop - exclusive
    - step_size = 1 (By default)

In [22]:
arr=np.arange(1,20,1)
arr

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19])

In [4]:
arr1=np.arange(1,10,2)
arr1

array([1, 3, 5, 7, 9])

In [23]:
arr=np.arange(1,11)
arr

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [24]:
arr=np.arange(11)
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [5]:
arr1=np.arange(12)
arr1

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [13]:
arr1=np.arange(5,len(arr1))
arr1

array([], dtype=int32)

In [7]:
lst=[i for i in range(1,11)]
arr=np.arange(1,len(lst))
arr**2

array([ 1,  4,  9, 16, 25, 36, 49, 64, 81])

In [26]:
lst = [i for i in range(1,11)]
arr = np.arange(1,len(lst),dtype = 'object')
arr**3

array([1, 8, 27, 64, 125, 216, 343, 512, 729], dtype=object)

### np.linspace()
- It creates an array of equally spaced values within the range of values.
- Syntax -
    - np.linspace(start , stop , Number of values you want within it)
    - By default, in linspace both the start and stop values are included
    - By default, linspace will return the answer in float

In [18]:
arr = np.linspace(2,5,20)
arr

array([2.        , 2.15789474, 2.31578947, 2.47368421, 2.63157895,
       2.78947368, 2.94736842, 3.10526316, 3.26315789, 3.42105263,
       3.57894737, 3.73684211, 3.89473684, 4.05263158, 4.21052632,
       4.36842105, 4.52631579, 4.68421053, 4.84210526, 5.        ])

In [23]:
arr1=np.linspace(5,22,5)
arr1

array([ 5.  ,  9.25, 13.5 , 17.75, 22.  ])

In [28]:
arr = np.linspace(2,5,10)
arr

array([2.        , 2.33333333, 2.66666667, 3.        , 3.33333333,
       3.66666667, 4.        , 4.33333333, 4.66666667, 5.        ])

In [29]:
arr = np.linspace(1,22,50)
arr

array([ 1.        ,  1.42857143,  1.85714286,  2.28571429,  2.71428571,
        3.14285714,  3.57142857,  4.        ,  4.42857143,  4.85714286,
        5.28571429,  5.71428571,  6.14285714,  6.57142857,  7.        ,
        7.42857143,  7.85714286,  8.28571429,  8.71428571,  9.14285714,
        9.57142857, 10.        , 10.42857143, 10.85714286, 11.28571429,
       11.71428571, 12.14285714, 12.57142857, 13.        , 13.42857143,
       13.85714286, 14.28571429, 14.71428571, 15.14285714, 15.57142857,
       16.        , 16.42857143, 16.85714286, 17.28571429, 17.71428571,
       18.14285714, 18.57142857, 19.        , 19.42857143, 19.85714286,
       20.28571429, 20.71428571, 21.14285714, 21.57142857, 22.        ])

### np.zeros()
- It will return an array of the given shape with all the elements of the array as zero.
- Syntax
   - np.zeros((rows,columns))
   - By default it will return object in float datatype

In [30]:
np.zeros((3,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [31]:
np.zeros((3,4))

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [33]:
arr = np.zeros((6, 6))
print(arr)

[[0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]]


In [35]:
np.zeros((4,3,6))


array([[[0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.]]])

In [34]:
np.zeros((3,4,6), dtype = 'int')

array([[[0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0]],

       [[0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0]],

       [[0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0]]])

### np.ones()
- It will return an array of the given shape with all the elements of the array as 1
- Syntax
   - np.ones((rows , columns))
 - By default it will return the object in float datatype.

In [35]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [36]:
arr = np.ones((6, ))
print(arr)

[1. 1. 1. 1. 1. 1.]


In [37]:
arr1 = np.ones((6, 5), dtype = int)
print(arr1)

[[1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]]


In [38]:
np.ones((3,4,6), dtype ='int')

array([[[1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1]],

       [[1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1]],

       [[1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1]]])

In [39]:
    np.ones((3,4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

### np.full()
- It will return an array of the given shape with all the elements of the array as value you have
entered.
- Syntax
   - np.full((rows , columns), values)
   - By default it will return the object in float datatype.

In [40]:
np.full((3,3) , 22)

array([[22, 22, 22],
       [22, 22, 22],
       [22, 22, 22]])

In [41]:
    np.full((3,4) ,"lavish", dtype = 'object')

array([['lavish', 'lavish', 'lavish', 'lavish'],
       ['lavish', 'lavish', 'lavish', 'lavish'],
       ['lavish', 'lavish', 'lavish', 'lavish']], dtype=object)

####  AS you can see in above code np.full is used for creating an specific numbers array as you can first we have to give rows and columns and then value we have to put it in.

In [42]:
np.full((3,4) , 2.2)

array([[2.2, 2.2, 2.2, 2.2],
       [2.2, 2.2, 2.2, 2.2],
       [2.2, 2.2, 2.2, 2.2]])

In [38]:
np.full((5,3),'sandhiya', dtype="object")

array([['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya']], dtype=object)

In [43]:
np.full((3,4) , 2+32j , dtype = 'object')

array([[(2+32j), (2+32j), (2+32j), (2+32j)],
       [(2+32j), (2+32j), (2+32j), (2+32j)],
       [(2+32j), (2+32j), (2+32j), (2+32j)]], dtype=object)

###  np.eye()
- It will return an identity matrix
- In an identity matrix, the diagonal elements are 1's and the non diagonal elements are 0's
- By default, the values will be in float
  - Syntax :
    - np.eye(rows , columns)

In [44]:
np.eye(4,5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.]])

In [39]:
np.eye(3,3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [45]:
np.eye(3,4,6, dtype= 'int')

array([[0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0]])

#### As you can see in upper code I've given 3 values it is giving all in 0's not giving diagonals of 1's in it.

In [41]:
np.eye(4,4,5, dtype= 'int')

array([[0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0]])

### np.zeros_like()
- It returns an array of 0's of the given shape of another array

In [45]:
arr = np.array([[1,2,3,4],[2,3,4,5],[5,6,7,8]])
arr

array([[1, 2, 3, 4],
       [2, 3, 4, 5],
       [5, 6, 7, 8]])

In [46]:
np.zeros_like(arr)

array([[0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0]])

###  np.full_like()
- it will return an array of the given shape with all the elements of the array as values you have entered.

In [47]:
arr

array([[1, 2, 3, 4],
       [2, 3, 4, 5],
       [5, 6, 7, 8]])

In [48]:
np.full_like(arr,22)

array([[22, 22, 22, 22],
       [22, 22, 22, 22],
       [22, 22, 22, 22]])

In [49]:
arr

array([[1, 2, 3, 4],
       [2, 3, 4, 5],
       [5, 6, 7, 8]])

In [50]:
np.full_like(arr,25)

array([[25, 25, 25, 25],
       [25, 25, 25, 25],
       [25, 25, 25, 25]])

###  np.diagonal()
- it return an array diagonal principal from an array
- BY default, function value of offset = 0

In [53]:
arr = np.random.randint(1,122,25)
arr


array([ 71, 114,  78,  74,  16,  52,  48,  24, 105,   5,  66,  40,  85,
       109,  91,  78,  42,  91,  38,  57,  57,  52,  43, 116,  73])

In [54]:
arr = arr.reshape(5,5)
arr


array([[ 71, 114,  78,  74,  16],
       [ 52,  48,  24, 105,   5],
       [ 66,  40,  85, 109,  91],
       [ 78,  42,  91,  38,  57],
       [ 57,  52,  43, 116,  73]])

In [55]:
np.diagonal(arr)

array([71, 48, 85, 38, 73])

In [56]:
np.diagonal(arr, offset=1)

array([114,  24, 109,  57])

In [57]:
np.diagonal(arr, offset=-1)

array([ 52,  40,  91, 116])

In [58]:
np.diagonal(arr, offset=0)

array([71, 48, 85, 38, 73])

In [59]:
np.diagonal(arr, offset=2)

array([ 78, 105,  91])

### np.diag()
- It creates an diagonal array.
- In a diagonal array, the diagonal elements are non-zero rest every element is zero
- Syntax:-
   - np.diag([diagonal])

In [60]:
arr1=np.diag([1,2,3,4,5,6])
arr1

array([[1, 0, 0, 0, 0, 0],
       [0, 2, 0, 0, 0, 0],
       [0, 0, 3, 0, 0, 0],
       [0, 0, 0, 4, 0, 0],
       [0, 0, 0, 0, 5, 0],
       [0, 0, 0, 0, 0, 6]])

In [62]:
arr1.T

array([[1, 0, 0, 0, 0, 0],
       [0, 2, 0, 0, 0, 0],
       [0, 0, 3, 0, 0, 0],
       [0, 0, 0, 4, 0, 0],
       [0, 0, 0, 0, 5, 0],
       [0, 0, 0, 0, 0, 6]])

In [61]:
arr2 = np.diag([[1,2,3,4,5] , [6,7,89,8,9]])
arr2

array([1, 7])

In [67]:
arr2.T

array([1, 7])

In [68]:
arr3 = np.diag([[1,2,3] , [4,5,6] , [7,8,9]])
arr3

array([1, 5, 9])

In [69]:
arr3.T

array([1, 5, 9])

In [69]:
a = np.random.randint(1,20,16)
a

array([ 5, 12,  4,  9,  6, 18, 16, 16,  7, 13, 13,  5,  4,  8,  1,  4])

In [70]:
numpyyy = a.reshape(4,4)
numpyyy

array([[ 5, 12,  4,  9],
       [ 6, 18, 16, 16],
       [ 7, 13, 13,  5],
       [ 4,  8,  1,  4]])

In [71]:
np.diagonal(numpyyy)

array([ 5, 18, 13,  4])

In [72]:
print(np.diagonal(numpyyy, offset=0))
print(np.diagonal(numpyyy , offset=1))
print(np.diagonal(numpyyy , offset=-1))
print(np.diagonal(numpyyy , offset=2))
print(np.diagonal(numpyyy , offset=-2))

[ 5 18 13  4]
[12 16  5]
[ 6 13  1]
[ 4 16]
[7 8]


In [73]:
np.diag(numpyyy)

array([ 5, 18, 13,  4])

In [74]:
x = np.arange(9).reshape(3,3)
x

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [75]:
print(np.diag(x))
print(np.diag(x, k=1))
print(np.diag(x, k=-1))

[0 4 8]
[1 5]
[3 7]


###  np.trace()
- it gives the sum of diagonals of array.

In [76]:
numpyyy

array([[ 5, 12,  4,  9],
       [ 6, 18, 16, 16],
       [ 7, 13, 13,  5],
       [ 4,  8,  1,  4]])

In [77]:
np.trace(numpyyy) # 10+17+11+8=46

40

In [78]:
np.trace(numpyyy,offset=1)

33

In [79]:
#As you can see it gives values between 0 and 1 whichever value you entered.
np.random.rand(2)

array([0.87133995, 0.306108  ])

In [80]:
np.random.randn(21)

array([-1.69414702,  0.07196977, -1.29244494,  0.34368481,  0.96078525,
        0.58353268,  0.25872901, -1.2981912 ,  0.96407929, -0.48885233,
       -0.49503801, -0.58153507, -0.35013631, -1.13188034,  0.68030172,
        0.7927653 ,  1.76306089,  1.65029497,  2.06093593,  1.89024646,
        1.24949858])

### Attributes Description
- ndim - returns number of dimension of the array
- size - returns number of elements in the array
- dtype - returns data type of elements in the array
- shape - returns the size of the array in each dimension.
- itemsize - returns the size (in bytes) of each elements in the array
- data - returns the buffer containing actual elements of the array in memory

### shape()
- This tell us about the shape of array.
- It tells us about the columns and rows. first it will tell about the columns then rows.

In [81]:
arr1=np.random.randint(1,22,20)
arr1

array([ 4,  8,  5, 19, 19, 20, 10, 16,  7, 12, 19,  7, 16, 20, 10,  8, 15,
        4,  3, 12])

In [82]:
arr1.shape

(20,)

In [83]:
numpyyy

array([[ 5, 12,  4,  9],
       [ 6, 18, 16, 16],
       [ 7, 13, 13,  5],
       [ 4,  8,  1,  4]])

In [85]:
numpyyy.ndim

2

In [86]:
numpyyy.shape

(4, 4)

In [87]:
numpyyy.dtype

dtype('int32')

In [86]:
numpyyy.size

16

In [88]:
a

array([ 5, 12,  4,  9,  6, 18, 16, 16,  7, 13, 13,  5,  4,  8,  1,  4])

In [89]:
print(a.shape)
print(a.itemsize)
print(a.data)

(16,)
4


In [89]:
numpyyy.itemsize

4

In [90]:
numpyyy.data

### Reshape
- It reshapes the array as per the shape we have passed
- It return a new array of the given shape, it doesnot make the changes in the existing array

In [91]:
a

array([10, 12, 11,  6,  5, 17, 15, 16,  1,  7, 11, 13, 13,  6, 11,  8])

In [92]:
a.reshape(2,2,4)

array([[[10, 12, 11,  6],
        [ 5, 17, 15, 16]],

       [[ 1,  7, 11, 13],
        [13,  6, 11,  8]]])

In [94]:
a.reshape(4,2,2)

array([[[10, 12],
        [11,  6]],

       [[ 5, 17],
        [15, 16]],

       [[ 1,  7],
        [11, 13]],

       [[13,  6],
        [11,  8]]])

In [95]:
a.itemsize

4

In [96]:
np.size(a)

16

#### In this upper code (2,2,4) and (4,2,2) reliee on this concept
- number of layers, list in each layer, elements in each list

###  dtype()
- dtype tells us about what type of data in the array.

In [97]:
numpyyy

array([[10, 12, 11,  6],
       [ 5, 17, 15, 16],
       [ 1,  7, 11, 13],
       [13,  6, 11,  8]])

In [98]:
numpyyy.dtype

dtype('int32')

In [102]:
b=np.full((4,3), "sandhiya")
b

array([['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya']], dtype='<U8')

In [103]:
b.dtype

dtype('<U8')

In [105]:
b=np.full((4,3), "sandhiya", dtype='object')
b

array([['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya']], dtype=object)

In [107]:
b.dtype

dtype('O')

### ndim() - Dimensionality
- It tell us about the dimensions of array.

In [108]:
a

array([10, 12, 11,  6,  5, 17, 15, 16,  1,  7, 11, 13, 13,  6, 11,  8])

In [109]:
a.ndim

1

In [112]:
numpyyy.ndim

2

In [113]:
b

array([['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya'],
       ['sandhiya', 'sandhiya', 'sandhiya']], dtype=object)

In [115]:
arr1=np.array([[[1,2,3],[4,5,6],[7,8,9],[5,6,7],[7,8,9]]])
arr1

array([[[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9],
        [5, 6, 7],
        [7, 8, 9]]])

In [116]:
arr1.ndim

3

In [117]:
num1 = np.random.randint(1,40,24).reshape(2,3,4)
num1

array([[[34, 25,  8, 11],
        [39, 37,  7, 32],
        [11,  8, 13, 20]],

       [[31, 12,  7, 22],
        [23,  4,  4,  8],
        [12,  9, 32, 21]]])

In [118]:
num1.ndim

3

In [119]:
num2 = np.random.randint(1,40,36).reshape(2,3,3,2)
num2

array([[[[37, 26],
         [33, 10],
         [28, 24]],

        [[ 1, 15],
         [22,  9],
         [27, 35]],

        [[33, 27],
         [18, 39],
         [23, 15]]],


       [[[ 7, 15],
         [39, 13],
         [15, 22]],

        [[ 5, 27],
         [24, 38],
         [26, 37]],

        [[18, 19],
         [34,  1],
         [25, 31]]]])

In [120]:
num2.ndim

4

### How to Create a 2D array and 3D array

In [121]:
arr=np.array([[1,2,3,4,5],[6,7,8,9,1]])
arr

array([[1, 2, 3, 4, 5],
       [6, 7, 8, 9, 1]])

In [122]:
arr.ndim

2

### Indexing and Slicing in Array
#### Indexing and Slicing in 1-D array
- the indexing and the slicing of a 1-D array is similar to the indexing and the slicing of a list
- Array will support both Positive and negative indexing
#### Indexing and Slicing in 2-D array
- How many list are there in an array and how many elements are there in each list
- How many list out of the total we want and how many elements from the total we want
#### Indexing
- Fetching the elements out form a particular position or index
  -  Syntax:
      - arr[index]

In [91]:
numpyyy

array([[ 5, 12,  4,  9],
       [ 6, 18, 16, 16],
       [ 7, 13, 13,  5],
       [ 4,  8,  1,  4]])

In [97]:
print(numpyyy[0])
print(numpyyy[3])
print(numpyyy[1])
print(numpyyy[-1])
print(numpyyy[-2])
print(numpyyy[0][0])
print(numpyyy[-1][-1])
print(numpyyy[len(numpyyy)-2])
print(numpyyy[len(numpyyy)-1])
print(numpyyy[len(numpyyy)-3][-2])
print(numpyyy[len(numpyyy)-1][-2])
print(numpyyy[2][3])
print(numpyyy[2][0])

[ 5 12  4  9]
[4 8 1 4]
[ 6 18 16 16]
[4 8 1 4]
[ 7 13 13  5]
5
4
[ 7 13 13  5]
[4 8 1 4]
16
1
5
7


In [98]:
numpyyy[2,0]=55
numpyyy

array([[ 5, 12,  4,  9],
       [ 6, 18, 16, 16],
       [55, 13, 13,  5],
       [ 4,  8,  1,  4]])

In [107]:
np.random.seed(3)
arr1 = np.random.randint(1,50,20)
arr1

array([43, 25,  4,  9,  1, 22, 20, 11, 44, 42, 11, 22, 39, 33, 21, 45, 30,
       40, 15, 27])

In [109]:
arr = arr1.reshape(4,5)
arr

array([[43, 25,  4,  9,  1],
       [22, 20, 11, 44, 42],
       [11, 22, 39, 33, 21],
       [45, 30, 40, 15, 27]])

In [113]:
arr.ndim

2

In [116]:
len(arr)

4

In [117]:
arr[1,4]

42

### Slicing :
#### rows Slicing :
- variable[row_startingindex : row_endingindex : step_size]
#### Columns Slicing :
- variable[column_startindex : column_endingindex : step_size]
#### Rows and columns Slicing :
- variable[row_startingindex : row_endingindex : step_size ,column_startindex :
column_endingindex : step_size]
- startingindex = 0 (By default)
- endingindex = Exclusive (len(variable)+1)
-step_size = 0

In [118]:
arr

array([[43, 25,  4,  9,  1],
       [22, 20, 11, 44, 42],
       [11, 22, 39, 33, 21],
       [45, 30, 40, 15, 27]])

In [164]:
print(arr[0])
print("----------------------")
print(arr[-1])
print("----------------------")
print(arr[::1])
print("----------------------")
print(arr[::-1])

[43 25  4  9  1]
----------------------
[45 30 40 15 27]
----------------------
[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]
 [45 30 40 15 27]]
----------------------
[[45 30 40 15 27]
 [11 22 39 33 21]
 [22 20 11 44 42]
 [43 25  4  9  1]]


In [165]:
print(arr[0:3:])
print("----------------------")
print(arr[:len(arr):])

[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]]
----------------------
[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]
 [45 30 40 15 27]]


In [166]:
print(arr[:])
print("----------------------")
print(arr[1:3,:4])

[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]
 [45 30 40 15 27]]
----------------------
[[22 20 11 44]
 [11 22 39 33]]


In [167]:
arr

array([[43, 25,  4,  9,  1],
       [22, 20, 11, 44, 42],
       [11, 22, 39, 33, 21],
       [45, 30, 40, 15, 27]])

In [168]:
print(arr[:2:,:5:])
print("----------------------")
print(arr[:3:,:2:])

[[43 25  4  9  1]
 [22 20 11 44 42]]
----------------------
[[43 25]
 [22 20]
 [11 22]]


In [169]:
print(arr[::,::])
print("----------------------")
print(arr[2:4:,2:-1:])

[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]
 [45 30 40 15 27]]
----------------------
[[39 33]
 [40 15]]


In [170]:
print(arr[::,::])
print("----------------------")
print(arr[2::,:2:])

[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]
 [45 30 40 15 27]]
----------------------
[[11 22]
 [45 30]]


In [171]:
print(arr[::,::])
print("----------------------")
print(arr[1::,2:4:])

[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]
 [45 30 40 15 27]]
----------------------
[[11 44]
 [39 33]
 [40 15]]


In [172]:
print(arr[::,::])
print("----------------------")
print(arr[::-1,5:len(arr):-1])

[[43 25  4  9  1]
 [22 20 11 44 42]
 [11 22 39 33 21]
 [45 30 40 15 27]]
----------------------
[]


In [173]:
np.random.seed(1)
arr2 = np.random.randint(22,222,81).reshape(9,9)
arr2

array([[ 59, 162,  94, 159, 155, 101, 214, 166, 151],
       [ 93, 156,  47, 200,  42, 123, 168, 161, 178],
       [179, 164,  72,  90, 118, 108, 163, 159,  29],
       [ 85,  83,  44,  79,  23, 150,  82,  30, 163],
       [137, 197, 143,  52,  93, 153, 220, 171,  71],
       [ 79,  25, 218,  46,  65,  98,  48,  74, 102],
       [131, 137,  63,  37,  86, 218,  47, 133, 157],
       [ 48, 175, 126,  44,  31, 217, 148,  45, 147],
       [122, 177, 187,  79, 105, 188, 158,  54, 184]])

In [174]:
print(arr2[::,::])
print("----------------------")
print(arr2[:: ,8:len(arr2):])

[[ 59 162  94 159 155 101 214 166 151]
 [ 93 156  47 200  42 123 168 161 178]
 [179 164  72  90 118 108 163 159  29]
 [ 85  83  44  79  23 150  82  30 163]
 [137 197 143  52  93 153 220 171  71]
 [ 79  25 218  46  65  98  48  74 102]
 [131 137  63  37  86 218  47 133 157]
 [ 48 175 126  44  31 217 148  45 147]
 [122 177 187  79 105 188 158  54 184]]
----------------------
[[151]
 [178]
 [ 29]
 [163]
 [ 71]
 [102]
 [157]
 [147]
 [184]]


In [175]:
print(arr2[::,::])
print("--------------------------------------------------------")
print(arr2[::-1,8:len(arr2):])

[[ 59 162  94 159 155 101 214 166 151]
 [ 93 156  47 200  42 123 168 161 178]
 [179 164  72  90 118 108 163 159  29]
 [ 85  83  44  79  23 150  82  30 163]
 [137 197 143  52  93 153 220 171  71]
 [ 79  25 218  46  65  98  48  74 102]
 [131 137  63  37  86 218  47 133 157]
 [ 48 175 126  44  31 217 148  45 147]
 [122 177 187  79 105 188 158  54 184]]
--------------------------------------------------------
[[184]
 [147]
 [157]
 [102]
 [ 71]
 [163]
 [ 29]
 [178]
 [151]]


In [183]:
print(arr2[::,::])
print("----------------------------------------------")
print(arr2[-1::,3:4:])

[[ 59 162  94 159 155 101 214 166 151]
 [ 93 156  47 200  42 123 168 161 178]
 [179 164  72  90 118 108 163 159  29]
 [ 85  83  44  79  23 150  82  30 163]
 [137 197 143  52  93 153 220 171  71]
 [ 79  25 218  46  65  98  48  74 102]
 [131 137  63  37  86 218  47 133 157]
 [ 48 175 126  44  31 217 148  45 147]
 [122 177 187  79 105 188 158  54 184]]
----------------------------------------------
[[79]]


In [186]:
print(arr2[::,::])
print("----------------------------------------------")
print(arr2[::,6:8:])

[[ 59 162  94 159 155 101 214 166 151]
 [ 93 156  47 200  42 123 168 161 178]
 [179 164  72  90 118 108 163 159  29]
 [ 85  83  44  79  23 150  82  30 163]
 [137 197 143  52  93 153 220 171  71]
 [ 79  25 218  46  65  98  48  74 102]
 [131 137  63  37  86 218  47 133 157]
 [ 48 175 126  44  31 217 148  45 147]
 [122 177 187  79 105 188 158  54 184]]
----------------------------------------------
[[214 166]
 [168 161]
 [163 159]
 [ 82  30]
 [220 171]
 [ 48  74]
 [ 47 133]
 [148  45]
 [158  54]]


In [187]:
print(arr2[::,::])
print("----------------------------------------------")
print(np.diagonal(arr2)[::])
print("----------------------------------------------")
print(np.diagonal(arr2)[::-1])
print("----------------------------------------------")
print(np.diagonal(arr2, offset=1)[::])
print("----------------------------------------------")
print(np.diagonal(arr2, offset=-1)[::])
print("----------------------------------------------")
print(np.diag(arr2 , k=6)[::])
print("----------------------------------------------")
print(np.diagonal(arr2 , offset=8)[::])

[[ 59 162  94 159 155 101 214 166 151]
 [ 93 156  47 200  42 123 168 161 178]
 [179 164  72  90 118 108 163 159  29]
 [ 85  83  44  79  23 150  82  30 163]
 [137 197 143  52  93 153 220 171  71]
 [ 79  25 218  46  65  98  48  74 102]
 [131 137  63  37  86 218  47 133 157]
 [ 48 175 126  44  31 217 148  45 147]
 [122 177 187  79 105 188 158  54 184]]
----------------------------------------------
[ 59 156  72  79  93  98  47  45 184]
----------------------------------------------
[184  45  47  98  93  79  72 156  59]
----------------------------------------------
[162  47  90  23 153  48 133 147]
----------------------------------------------
[ 93 164  44  52  65 218 148  54]
----------------------------------------------
[214 161  29]
----------------------------------------------
[151]


#### You're on the right track – np.diagonals is well suited for this, but you need to take care computing the offsets. In numpy (and in matrix math by convention) the diagonals run {{topleft to bottom-right.}} The main diagonal has offset 0; the diagonals above this have positive offsets, and the diagonals below have negative offsets.

In [188]:
arr2    

array([[ 59, 162,  94, 159, 155, 101, 214, 166, 151],
       [ 93, 156,  47, 200,  42, 123, 168, 161, 178],
       [179, 164,  72,  90, 118, 108, 163, 159,  29],
       [ 85,  83,  44,  79,  23, 150,  82,  30, 163],
       [137, 197, 143,  52,  93, 153, 220, 171,  71],
       [ 79,  25, 218,  46,  65,  98,  48,  74, 102],
       [131, 137,  63,  37,  86, 218,  47, 133, 157],
       [ 48, 175, 126,  44,  31, 217, 148,  45, 147],
       [122, 177, 187,  79, 105, 188, 158,  54, 184]])

In [189]:
np.diag(arr2)

array([ 59, 156,  72,  79,  93,  98,  47,  45, 184])

In [190]:
np.diag(np.diag(arr2))

array([[ 59,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0, 156,   0,   0,   0,   0,   0,   0,   0],
       [  0,   0,  72,   0,   0,   0,   0,   0,   0],
       [  0,   0,   0,  79,   0,   0,   0,   0,   0],
       [  0,   0,   0,   0,  93,   0,   0,   0,   0],
       [  0,   0,   0,   0,   0,  98,   0,   0,   0],
       [  0,   0,   0,   0,   0,   0,  47,   0,   0],
       [  0,   0,   0,   0,   0,   0,   0,  45,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0, 184]])

In [191]:
arr1 = np.random.randint(30,70,121).reshape(11,11)
arr1

array([[40, 53, 45, 53, 55, 37, 58, 40, 62, 54, 53],
       [31, 43, 33, 30, 43, 36, 51, 36, 32, 42, 57],
       [51, 41, 37, 43, 38, 41, 42, 50, 60, 66, 69],
       [37, 34, 48, 62, 43, 40, 53, 47, 37, 54, 40],
       [58, 50, 62, 42, 31, 60, 54, 48, 63, 32, 58],
       [69, 64, 40, 62, 48, 52, 36, 32, 37, 69, 45],
       [35, 47, 50, 52, 67, 43, 63, 50, 40, 62, 55],
       [51, 69, 62, 49, 55, 44, 43, 50, 36, 43, 45],
       [54, 39, 32, 37, 35, 66, 67, 51, 43, 57, 43],
       [44, 53, 46, 58, 33, 32, 44, 30, 59, 41, 65],
       [59, 63, 32, 50, 49, 62, 58, 48, 61, 58, 60]])

In [192]:
np.diag(np.diag(arr1))

array([[40,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 43,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 37,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 62,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 31,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 52,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0, 63,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 50,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0, 43,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0, 41,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 60]])

In [193]:
arr1

array([[40, 53, 45, 53, 55, 37, 58, 40, 62, 54, 53],
       [31, 43, 33, 30, 43, 36, 51, 36, 32, 42, 57],
       [51, 41, 37, 43, 38, 41, 42, 50, 60, 66, 69],
       [37, 34, 48, 62, 43, 40, 53, 47, 37, 54, 40],
       [58, 50, 62, 42, 31, 60, 54, 48, 63, 32, 58],
       [69, 64, 40, 62, 48, 52, 36, 32, 37, 69, 45],
       [35, 47, 50, 52, 67, 43, 63, 50, 40, 62, 55],
       [51, 69, 62, 49, 55, 44, 43, 50, 36, 43, 45],
       [54, 39, 32, 37, 35, 66, 67, 51, 43, 57, 43],
       [44, 53, 46, 58, 33, 32, 44, 30, 59, 41, 65],
       [59, 63, 32, 50, 49, 62, 58, 48, 61, 58, 60]])

In [194]:
np.diagflat(np.diagflat(arr1))

array([[40,  0,  0, ...,  0,  0,  0],
       [ 0,  0,  0, ...,  0,  0,  0],
       [ 0,  0,  0, ...,  0,  0,  0],
       ...,
       [ 0,  0,  0, ...,  0,  0,  0],
       [ 0,  0,  0, ...,  0,  0,  0],
       [ 0,  0,  0, ...,  0,  0, 60]])

### diagflat() :
- It creates the diagonals from given array and rest of non diagonal values becomes 0's.

In [197]:
numm=np.diagflat([[22,23,2,4],[11,13,12,1]])
numm

array([[22,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 23,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  2,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  4,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 11,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 13,  0,  0],
       [ 0,  0,  0,  0,  0,  0, 12,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1]])

In [198]:
numm.ndim

2

In [202]:
x = np.diagflat([[[12,12],[22,22],
[33,33],[44,44]]])
x

array([[12,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 12,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 22,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 22,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 33,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 33,  0,  0],
       [ 0,  0,  0,  0,  0,  0, 44,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 44]])

In [203]:
x=np.diagonal([[[12,12],[22,22],
                [33,33],[44,44]]])
x

array([[12],
       [12]])

#### As you can see in above codes np.diag and np.diagonal doesn't create a array by giving arrays it is only applicable in np.diagflat

In [204]:
arr = np.random.random(9).reshape(3,3)
arr

array([[0.12134346, 0.04455188, 0.10749413],
       [0.22570934, 0.71298898, 0.55971698],
       [0.01255598, 0.07197428, 0.96727633]])

In [206]:
arr1 = np.random.rand(3,4)
arr1

array([[0.30594132, 0.74537515, 0.38821925, 0.46966765],
       [0.71004438, 0.36512663, 0.92725033, 0.73218711],
       [0.24383709, 0.67966439, 0.0595789 , 0.85628818]])

In [205]:
np.random.seed(6)
arr3 = np.random.randint(1,200,64).reshape(8,8)
arr3

array([[139, 107, 110,  80, 107,  81,  63,  26],
       [  2,  76,  78, 186,  27, 162, 111,  69],
       [162,   9, 131,  32, 195, 134, 126, 155],
       [ 16,  87, 115, 172, 198, 178, 133,  63],
       [ 81, 195, 128,  67,  64, 148,  86, 125],
       [ 29,  44,  58, 134,  51,  27,  79, 120],
       [ 41, 198,  71,  64,  11,  68,  86, 140],
       [130,  83, 164, 151,  42, 124, 106,  25]])

### Indexing and Slicing in 3-D Array
#### Indexing :
- Syntax :
    - arr[arr_index] --- It fetches the layer
    - arr[arr_index , arr_row] ---- It fetches the layer and the list
    - arr[arr_index , arr_row , arr_column] ---- It fetches layer and thelist and the elements inside lists.

In [207]:
import numpy as np

In [220]:
np.random.seed(1)
arr1 = np.random.randint(30,70,48).reshape(3,4,4)
arr1

array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [46, 31, 42, 37],
        [36, 55, 50, 67]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 62, 52],
        [43, 39, 37, 52]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [37, 33, 36, 51],
        [33, 34, 54, 42]]])

In [221]:
arr1[1]

array([[48, 50, 41, 58],
       [59, 44, 34, 53],
       [53, 60, 62, 52],
       [43, 39, 37, 52]])

In [222]:

arr1[0,2] = 22
arr1

array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 22, 22, 22],
        [36, 55, 50, 67]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 62, 52],
        [43, 39, 37, 52]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [37, 33, 36, 51],
        [33, 34, 54, 42]]])

In [227]:
arr1[0,2,(1,2)] = 11,12
arr1

array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 11, 12, 22],
        [36, 55, 50, 67]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 62, 52],
        [43, 39, 37, 52]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [37, 33, 36, 51],
        [33, 34, 54, 42]]])

In [231]:
arr1[2,2,(0,3)] = 73,65
arr1

array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 11, 12, 22],
        [36, 55, 50, 67]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 62, 52],
        [43, 39, 37, 52]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [73, 33, 36, 65],
        [33, 34, 54, 42]]])

In [232]:
arr1[1,2,(2,3)] = 26,25
arr1

array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 11, 12, 22],
        [36, 55, 50, 67]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 26, 25],
        [43, 39, 37, 52]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [73, 33, 36, 65],
        [33, 34, 54, 42]]])

### Slicing :
##### Syntax :
   - arr[layer_startindex : layer_endindex : step_size]
- layer_startindex = 0
- layer_endindex = Exclusive (by default)
- step size = 1 (By default)

In [233]:
arr1

array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 11, 12, 22],
        [36, 55, 50, 67]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 26, 25],
        [43, 39, 37, 52]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [73, 33, 36, 65],
        [33, 34, 54, 42]]])

In [237]:
arr1[:len(arr1):2]

array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 11, 12, 22],
        [36, 55, 50, 67]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [73, 33, 36, 65],
        [33, 34, 54, 42]]])

In [238]:
arr1[::-1]

array([[[31, 30, 47, 38],
        [54, 43, 38, 60],
        [73, 33, 36, 65],
        [33, 34, 54, 42]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 26, 25],
        [43, 39, 37, 52]],

       [[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 11, 12, 22],
        [36, 55, 50, 67]]])

### Now,
 - arr[layer_startindex : layer_endindex : step_size ,row_startindex : row_endindex : step_size]
- startindex = 0
- endindex = Exclusive (by default)
- step size = 1 (By default)

In [239]:
arr1


array([[[67, 42, 38, 39],
        [41, 35, 45, 30],
        [22, 11, 12, 22],
        [36, 55, 50, 67]],

       [[48, 50, 41, 58],
        [59, 44, 34, 53],
        [53, 60, 26, 25],
        [43, 39, 37, 52]],

       [[31, 30, 47, 38],
        [54, 43, 38, 60],
        [73, 33, 36, 65],
        [33, 34, 54, 42]]])

In [240]:
print(arr1[:: , :: , 3::])

[[[39]
  [30]
  [22]
  [67]]

 [[58]
  [53]
  [25]
  [52]]

 [[38]
  [60]
  [65]
  [42]]]


In [241]:
print(arr1[::2 , :2: , 3::])

[[[39]
  [30]]

 [[38]
  [60]]]


In [242]:
print(arr1[:2: , 2:: , 1:3:])

[[[11 12]
  [55 50]]

 [[60 26]
  [39 37]]]


In [244]:
np.random.seed(1)
varr = np.random.randint(22,172,49).reshape(7,7)
varr

array([[ 59, 162,  94, 159, 155, 101, 166],
       [151,  93, 156,  47,  42, 123, 168],
       [161, 164,  72,  90, 118, 108, 163],
       [159,  29,  85,  83,  44,  79,  23],
       [150,  82,  30, 163, 137, 143,  52],
       [ 93, 153, 171,  71,  79,  25,  46],
       [ 65,  98,  48,  74, 102, 131, 137]])

In [245]:
np.diag(varr)[::-1]

array([137,  25, 137,  83,  72,  93,  59])

In [246]:
np.diag(varr)

array([ 59,  93,  72,  83, 137,  25, 137])

##### Now Now Now, You can see in Below code that in diagonal it is possible that we can extract the diagonal from {right-top-left-bottom}

In [247]:
np.diag(varr[::-1])[::]

array([ 65, 153,  30,  83, 118, 123, 166])

In [248]:
np.diag(varr[::-1])[::-1]

array([166, 123, 118,  83,  30, 153,  65])

### Array of Random Value
- In numpy, we have a random module that is used to create array of random values in it random.random()
- It generates a specified number of random values
- The random valuws are always between 0 and 1
##### Syntax:
- np.random.random(number of values)

In [249]:
x = np.random.random(24)
x

array([0.75014431, 0.98886109, 0.74816565, 0.28044399, 0.78927933,
       0.10322601, 0.44789353, 0.9085955 , 0.29361415, 0.28777534,
       0.13002857, 0.01936696, 0.67883553, 0.21162812, 0.26554666,
       0.49157316, 0.05336255, 0.57411761, 0.14672857, 0.58930554,
       0.69975836, 0.10233443, 0.41405599, 0.69440016])

In [250]:
i = np.random.random(36).reshape(3,4,3)
i

array([[[0.41417927, 0.04995346, 0.53589641],
        [0.66379465, 0.51488911, 0.94459476],
        [0.58655504, 0.90340192, 0.1374747 ],
        [0.13927635, 0.80739129, 0.39767684]],

       [[0.1653542 , 0.92750858, 0.34776586],
        [0.7508121 , 0.72599799, 0.88330609],
        [0.62367221, 0.75094243, 0.34889834],
        [0.26992789, 0.89588622, 0.42809119]],

       [[0.96484005, 0.6634415 , 0.62169572],
        [0.11474597, 0.94948926, 0.44991213],
        [0.57838961, 0.4081368 , 0.23702698],
        [0.90337952, 0.57367949, 0.00287033]]])

#### random.randint()
- It returns a 1-D array of random numbers (integers) between a specified range
##### Synatx:
- np.random.randint(start, stop, number of values)
- By default, both the start and the stop values are excluded

In [251]:
v = np.random.randint(1,22,16).reshape(4,4)
v

array([[18, 11, 21, 17],
       [ 6, 14, 18, 19],
       [ 2, 21, 11,  1],
       [ 8,  1, 20, 18]])

### random.seed(random state)
- It is used to set the random state of the randomized value
- Random state is an integer number in the range 0 to (2**32)-1 

In [252]:
np.random.seed(10)
arrr = np.random.randint(0,51,36).reshape(6,6)
arrr

array([[ 9, 36, 15,  0, 49, 28],
       [25, 29, 48, 29, 49,  8],
       [ 9,  0, 42, 40, 36, 16],
       [36, 47, 11, 24, 43, 33],
       [ 8, 36, 14, 49, 13,  5],
       [13, 25, 13, 28, 22, 30]])

In [253]:
np.random.seed(10)
arr1 = np.random.rand(6,6)
arr1

array([[0.77132064, 0.02075195, 0.63364823, 0.74880388, 0.49850701,
        0.22479665],
       [0.19806286, 0.76053071, 0.16911084, 0.08833981, 0.68535982,
        0.95339335],
       [0.00394827, 0.51219226, 0.81262096, 0.61252607, 0.72175532,
        0.29187607],
       [0.91777412, 0.71457578, 0.54254437, 0.14217005, 0.37334076,
        0.67413362],
       [0.44183317, 0.43401399, 0.61776698, 0.51313824, 0.65039718,
        0.60103895],
       [0.8052232 , 0.52164715, 0.90864888, 0.31923609, 0.09045935,
        0.30070006]])

In [254]:
(2**32)-1

4294967295

### random.normal()
- It returns an array of random value those are normally distributed
- Syntax: np.random.normal(mean, sd, size) 

In [255]:
varry = np.random.normal(22, 4 , 50)
varry

array([22.53083318, 20.09543194, 27.23389232, 22.78005312, 23.60083995,
       20.64947065, 27.02588906, 19.07212199, 24.6409262 , 20.59651243,
       18.24226656, 20.04265113, 18.78163543, 21.14920945, 20.64343902,
       23.24867975, 24.26061068, 21.41031897, 21.89637865, 23.15637682,
       19.84048371, 24.83264008, 25.36889895, 22.81432319, 31.57881466,
       25.66983575, 21.55091012, 20.55127821, 21.07127098, 19.9930844 ,
       26.51514061, 19.20875988, 21.67551126, 19.88281568, 26.18473143,
       16.32577588, 20.55000327, 21.51237724, 23.27742568, 23.84361161,
       21.13684043, 25.95628983, 23.25901512, 31.87060423, 15.96671405,
       24.48240265, 17.81946985, 18.80796473, 29.94033837, 28.97925659])

### Functions in Array
#### np.sum()
- It returns the sum of the array.
##### Syntax :
- np.sum(arr , axis = 0)
- arr.sum(axis = 0,1)
    - axis = 0 ------> Interpreter reads the array from top to bottom which known as row wise operation.
     - axis = 1 ------> Interpreter reads the array from left to right which known as column wise operation.

In [256]:
np.random.seed(1)
axxxiiss = np.random.randint(1,22,14)
axxxiiss

array([ 6, 12, 13,  9, 10, 12,  6, 16,  1, 17,  2, 13,  8, 14])

In [257]:
np.sum(axxxiiss, axis=0)

139

In [258]:
axxxiiss.sum()

139

In [259]:
axxxiiss.sum(axis = 1)

AxisError: axis 1 is out of bounds for array of dimension 1

##### As you can see in upper code is giving error by out of bound for array of 1-D , axis = 0,1 is applicable in 2-D , 3-D

In [260]:
np.random.seed(1)
i_ = np.random.randint(0,51,16).reshape(4,4)
i_

array([[37, 43, 12,  8],
       [ 9, 11,  5, 15],
       [ 0, 16,  1, 12],
       [ 7, 45,  6, 25]])

In [261]:
np.sum(i_ , axis= 0)

array([ 53, 115,  24,  60])

In [262]:
i_.sum(axis = 1)

array([100,  40,  29,  83])

In [263]:
val = np.random.randint(22,52,16).reshape(2,2,4)
val

array([[[40, 42, 27, 40],
        [42, 33, 50, 32]],

       [[50, 51, 36, 40],
        [26, 45, 45, 31]]])

In [264]:
np.sum(val , axis=0)

array([[90, 93, 63, 80],
       [68, 78, 95, 63]])

### np.mean()
- It returns the average (mean) of the array
##### Syntax:
- np.mean(arr, axis = 0,1)
- arr.mean(axis = 0,1)

In [265]:
np.random.seed(1)
arr = np.random.randint(1,51,36).reshape(6,6)
arr

array([[38, 44, 13,  9, 10, 12],
       [ 6, 16,  1, 17,  2, 13],
       [ 8, 46,  7, 26, 21, 38],
       [19, 21, 12, 43, 29, 30],
       [15,  5, 24, 24, 42, 50],
       [31, 33, 23, 14, 42, 10]])

In [266]:
np.mean(arr) 

22.055555555555557

#### To Calculate mean,
- Formula : sum of array / no. of elements

In [267]:
print(np.mean(arr ,axis=0))
print("--------------------------")
(38+
6+
8+
19+
15+
31)/6
# no of rows = 6

[19.5        27.5        13.33333333 22.16666667 24.33333333 25.5       ]
--------------------------


19.5

In [268]:
print(np.mean(arr ,axis=1))
print("--------------------------")
(38 + 44 + 13 + 9 + 10 + 12)/6
# no of columns = 6

[21.          9.16666667 24.33333333 25.66666667 26.66666667 25.5       ]
--------------------------


21.0

### np.median()
- It return the median of an array
#### Syntax:
- np.median(arr, axis = 0,1)

In [269]:
arr

array([[38, 44, 13,  9, 10, 12],
       [ 6, 16,  1, 17,  2, 13],
       [ 8, 46,  7, 26, 21, 38],
       [19, 21, 12, 43, 29, 30],
       [15,  5, 24, 24, 42, 50],
       [31, 33, 23, 14, 42, 10]])

In [270]:
np.median(arr , axis = 0)

array([17. , 27. , 12.5, 20.5, 25. , 21.5])

In [271]:
print("sorted of rows axis 6 + 8 + 15 + 19 + 31 + 38 : ", (15 + 19)/2)
print("15 AND 19 are the sorted middle numbers of rows axis ")

sorted of rows axis 6 + 8 + 15 + 19 + 31 + 38 :  17.0
15 AND 19 are the sorted middle numbers of rows axis 


In [272]:
np.median(arr , axis = 1)

array([12.5,  9.5, 23.5, 25. , 24. , 27. ])

In [273]:
print("sorted of columns axis 9 + 10 + 12 + 13 + 38 + 44 : ", (12 + 13)/2)
print("12 AND 13 are the sorted middle numbers of columns axis ")

sorted of columns axis 9 + 10 + 12 + 13 + 38 + 44 :  12.5
12 AND 13 are the sorted middle numbers of columns axis 


#### Formula for median is : sorted middle indexes of (rows/columns) of an array /2

### np.var()
- It return the variance of the array
#### Syntax:
- np.var(arr, axis = 0,1)

In [275]:
np.random.seed(1)
x = np.random.randint(1,51,16).reshape(4,4)
x

array([[38, 44, 13,  9],
       [10, 12,  6, 16],
       [ 1, 17,  2, 13],
       [ 8, 46,  7, 26]])

In [277]:
x.var()

190.3125

In [278]:
np.var(x , axis=0)

array([199.1875, 236.1875,  15.5   ,  39.5   ])

In [279]:
np.mean(x , axis=0)

array([14.25, 29.75,  7.  , 16.  ])

In [280]:
np.mean(x)

16.75

In [281]:
np.var(x , axis=1)

array([231.5   ,  13.    ,  47.6875, 253.1875])

In [282]:
np.mean(x , axis=1)

array([26.  , 11.  ,  8.25, 21.75])

#### Formula : Number of axis(0 (rows),1(Columns)) - Mean of axis(0(rows) ,1(columns))**2 + That whole axis(0,1) / number of layers or array

In [283]:
((38-14.25)**2 + (10-14.25)**2 +(1-14.25)**2 +(8-14.25)**2)/4

199.1875

In [284]:
((38-26)**2 + (44-26)**2 +(13-26)**2 +(9-26)**2)/4

231.5

### np.std()
- It returns the standard deviation of an array
#### Syntax:
- np.std(arr, axis = 0,1)

In [285]:
np.random.seed(1)
axisss = np.random.randint(1,30,25).reshape(5,5)
axisss

array([[ 6, 12, 13,  9, 10],
       [12,  6, 16,  1, 17],
       [ 2, 13,  8, 14, 29],
       [ 7, 26, 19, 21,  6],
       [19, 21, 12, 29, 11]])

In [286]:
v = np.mean(axisss , axis=0)
c = np.var(axisss , axis=0)
print(v)
print(c)

[ 9.2 15.6 13.6 14.8 14.6]
[34.16 49.84 13.84 92.96 64.24]


#### Solving of Variances

In [287]:
((6-9.2)**2 + (12-9.2)**2 + (2-9.2)**2 + (7-9.2)**2 + (19-9.2)**2)/5

34.160000000000004

In [288]:
np.std(axisss , axis= 0)

array([5.84465568, 7.05974504, 3.72021505, 9.64157663, 8.01498596])

#### Formula : sqrt(variance of axis (0,1))

In [291]:
import math

In [292]:
math.sqrt(34.16)

5.844655678480983

#### Note : For finding Standard Deviation : 1. We have to first find the mean(np.mean()) of the array of axis(0,1)
#### 2. After that from mean we have to find the variance(np.var()) of the array of axis(0,1)
#### 3. Afterwards from that We have to do square of variances of that particular axis(0,1) for standard deviation(np.std()) we have to find.

In [293]:
np.random.seed(1)
garryy=np.random.randint(1,22,25).reshape(5,5)
garryy

array([[ 6, 12, 13,  9, 10],
       [12,  6, 16,  1, 17],
       [ 2, 13,  8, 14,  7],
       [19, 21,  6, 19, 21],
       [12, 11, 15, 19,  5]])

In [294]:
print(np.mean(garryy))
print(np.var(garryy))
print(np.std(garryy))

11.76
31.862399999999997
5.644678910265844


In [295]:
math.sqrt(31.8629999999997)

5.644732057414213

### np.min()
- It return the minimum value of the arary
#### Syntax:
- np.min(arr, axis = 0,1)

In [296]:
garryy

array([[ 6, 12, 13,  9, 10],
       [12,  6, 16,  1, 17],
       [ 2, 13,  8, 14,  7],
       [19, 21,  6, 19, 21],
       [12, 11, 15, 19,  5]])

In [297]:
np.min(garryy)

1

#### As you can see in below it shows that by finding np.min() we find minimum value and  checks the all the axis of 0 (rows) of the array and give each value from each row specifically minimum value.

In [298]:
np.min(garryy, axis=0)

array([2, 6, 6, 1, 5])

In [299]:
np.min(garryy, axis=1)

array([6, 1, 2, 6, 5])

### np.max()
- It return the maximum value of the arary
#### Syntax:
- np.max(arr, axis = 0,1)

In [300]:
garryy

array([[ 6, 12, 13,  9, 10],
       [12,  6, 16,  1, 17],
       [ 2, 13,  8, 14,  7],
       [19, 21,  6, 19, 21],
       [12, 11, 15, 19,  5]])

In [301]:
np.max(garryy)

21

#### As you can see in below it shows that by finding np.max() we find maximum value and it checks the all the axis of 0 (rows) of the array and give each value from each row specifically

In [302]:
np.max(garryy, axis=0)

array([19, 21, 16, 19, 21])

In [304]:
np.max(garryy, axis=1)

array([13, 17, 14, 21, 19])

### np.percentile()
- It return the specified percentile value of the array

In [305]:
garryy

array([[ 6, 12, 13,  9, 10],
       [12,  6, 16,  1, 17],
       [ 2, 13,  8, 14,  7],
       [19, 21,  6, 19, 21],
       [12, 11, 15, 19,  5]])

In [306]:
np.percentile(garryy,10)

5.4

In [310]:
len(garryy)

5

In [311]:
np.percentile(garryy , 7)

4.040000000000001

In [312]:
a = garryy.flatten()
a

array([ 6, 12, 13,  9, 10, 12,  6, 16,  1, 17,  2, 13,  8, 14,  7, 19, 21,
        6, 19, 21, 12, 11, 15, 19,  5])

In [313]:
np.sort(a)

array([ 1,  2,  5,  6,  6,  6,  7,  8,  9, 10, 11, 12, 12, 12, 13, 13, 14,
       15, 16, 17, 19, 19, 19, 21, 21])

### astype() :
- it typecast one daya type to another.

In [2]:
import numpy as np

In [3]:
arr1=np.array([2,3.8,8,89,77.3,66.5,33.2,12.3,45])
arr1

array([ 2. ,  3.8,  8. , 89. , 77.3, 66.5, 33.2, 12.3, 45. ])

In [4]:
arr1.astype(int)

array([ 2,  3,  8, 89, 77, 66, 33, 12, 45])

In [5]:
arr1.astype(complex)

array([ 2. +0.j,  3.8+0.j,  8. +0.j, 89. +0.j, 77.3+0.j, 66.5+0.j,
       33.2+0.j, 12.3+0.j, 45. +0.j])

In [6]:
arr1.astype(float)

array([ 2. ,  3.8,  8. , 89. , 77.3, 66.5, 33.2, 12.3, 45. ])

In [8]:
arr1.astype(str)

array(['2.0', '3.8', '8.0', '89.0', '77.3', '66.5', '33.2', '12.3',
       '45.0'], dtype='<U32')

In [9]:
arr1.astype(set)

array([2.0, 3.8, 8.0, 89.0, 77.3, 66.5, 33.2, 12.3, 45.0], dtype=object)

In [10]:
arr1.astype(list)

array([2.0, 3.8, 8.0, 89.0, 77.3, 66.5, 33.2, 12.3, 45.0], dtype=object)

In [11]:
arr1.astype(bool)

array([ True,  True,  True,  True,  True,  True,  True,  True,  True])

### Null Values in NUMPY
- In Python normal null value represent by the NONE
- But in Numpy library We can represent the null value by np.nan
- by default np.nan is 1 ----> which means that np.nan is true by default.
- NAN stands for Not A Number
- By default np.nan gives value in float datatype.

In [12]:
a=None
print(a,type(a))

None <class 'NoneType'>


In [13]:
arr1=np.array([np.nan])
print(arr1, type(arr1))

[nan] <class 'numpy.ndarray'>


In [14]:
arr1.dtype

dtype('float64')

In [15]:
arr2 = np.array([12, 34, np.nan , 32 , np.nan , 45 , 67, np.nan , 99 ]).reshape(3,3)
arr2

array([[12., 34., nan],
       [32., nan, 45.],
       [67., nan, 99.]])

#### As you can see now by default array is in float data type. because of np.nan

In [16]:
arr2.dtype

dtype('float64')

### Operations on Null Values
- Count the number of null values in array.
- Accessing the null values from an array.
- Count the non null values in the array.
- Accessing the non null values from an array.
- Replacing the null values in the array.
##### np.isnan()
- It checks for nan values and returns a boolean array, where True means the null value is
present and False means the non-null values is present

In [17]:
arr2

array([[12., 34., nan],
       [32., nan, 45.],
       [67., nan, 99.]])

In [18]:
np.isnan(arr2)

array([[False, False,  True],
       [False,  True, False],
       [False,  True, False]])

###  Count the null values and Non null values inside the array
##### Syntax :
- np.isnan().sum() ----- For Null values
- (~ np.isnan()).sum() ----- For non Null Values

In [19]:
arr2

array([[12., 34., nan],
       [32., nan, 45.],
       [67., nan, 99.]])

In [20]:
np.isnan(arr2)

array([[False, False,  True],
       [False,  True, False],
       [False,  True, False]])

In [21]:
np.isnan(arr2).sum()

3

In [22]:
(~np.isnan(arr2)).sum()

6

In [23]:
(np.isnan(arr2)== False).sum()

6

In [24]:
len(arr2[np.isnan(arr2) == False])

6

In [25]:
arr2[np.isnan(arr2) == True]

array([nan, nan, nan])

In [26]:
arr2[(~np.isnan(arr2) == True)]

array([12., 34., 32., 45., 67., 99.])

### Accessing the null values from an array
##### Syntax :
- np.where(condition)
- np.where(condition)
- It helps to fetch the values based on some conditions
- It returns the index of the element where the condition is True
- To fetch the element we pass np.where () as the index to the array object

In [27]:
arr2

array([[12., 34., nan],
       [32., nan, 45.],
       [67., nan, 99.]])

In [28]:
np.where(np.isnan(arr2))

(array([0, 1, 2], dtype=int64), array([2, 1, 1], dtype=int64))

#### As you can see in above code that it is returning two arrays first tarray tells us about the layers or lists in the array by indexing numbers and the another array states the position where nan values are there.
#### As you can see at 0th list --- at 2nd position there is an nan value.

In [30]:
np.random.seed(1)
arr3=np.random.randint(1,51,36).reshape(4,9)
arr3

array([[38, 44, 13,  9, 10, 12,  6, 16,  1],
       [17,  2, 13,  8, 46,  7, 26, 21, 38],
       [19, 21, 12, 43, 29, 30, 15,  5, 24],
       [24, 42, 50, 31, 33, 23, 14, 42, 10]])

In [31]:
np.where(arr3>35)

(array([0, 0, 1, 1, 2, 3, 3, 3], dtype=int64),
 array([0, 1, 4, 8, 3, 1, 2, 7], dtype=int64))

In [32]:
np.random.seed(45)
arr=np.random.randint(10,100,12).reshape(3,2,2)
arr

array([[[85, 40],
        [13, 42]],

       [[71, 95],
        [45, 78]],

       [[25, 75],
        [24, 63]]])

In [33]:
np.where(arr>40)

(array([0, 0, 1, 1, 1, 1, 2, 2], dtype=int64),
 array([0, 1, 0, 0, 1, 1, 0, 1], dtype=int64),
 array([0, 1, 0, 1, 0, 1, 1, 1], dtype=int64))

In [34]:
arr>50

array([[[ True, False],
        [False, False]],

       [[ True,  True],
        [False,  True]],

       [[False,  True],
        [False,  True]]])

In [53]:
arr7 = np.array([12, np.nan , np.nan , 55 , np.nan , 99 , np.nan , 76 , np.nan , 45, 345,1 ])
arr7

array([ 12.,  nan,  nan,  55.,  nan,  99.,  nan,  76.,  nan,  45., 345.,
         1.])

In [54]:
arr7[np.where(np.isnan(arr7))]

array([nan, nan, nan, nan, nan])

In [55]:
arr7[np.isnan(arr7)== True]

array([nan, nan, nan, nan, nan])

### Accessing the non Null Values

In [56]:
arr

array([[[85, 40],
        [13, 42]],

       [[71, 95],
        [45, 78]],

       [[25, 75],
        [24, 63]]])

In [57]:
arr[np.where(arr>20)]

array([85, 40, 42, 71, 95, 45, 78, 25, 75, 24, 63])

In [58]:
arr[(arr>20) == True]

array([85, 40, 42, 71, 95, 45, 78, 25, 75, 24, 63])

### np.where --- Case 2
- Syntax : np.where(condition, x, y)
- If the condition is True, then x is the result, otherwise y is the result

In [59]:
np.random.seed(10)
arr2=np.random.randint(1,51,36).reshape(6,6)
arr2

array([[10, 37, 16,  1, 50, 29],
       [26, 30, 49, 30, 50,  9],
       [10,  1, 43, 41, 37, 17],
       [37, 48, 12, 25, 44, 34],
       [ 9, 37, 15, 50, 14,  6],
       [14, 26, 14, 29, 23, 31]])

In [60]:
np.where(arr2%2==0, "Even","odd")

array([['Even', 'odd', 'Even', 'odd', 'Even', 'odd'],
       ['Even', 'Even', 'odd', 'Even', 'Even', 'odd'],
       ['Even', 'odd', 'odd', 'odd', 'odd', 'odd'],
       ['odd', 'Even', 'Even', 'odd', 'Even', 'Even'],
       ['odd', 'odd', 'odd', 'Even', 'Even', 'Even'],
       ['Even', 'Even', 'Even', 'odd', 'odd', 'odd']], dtype='<U4')

In [61]:
np.where(arr2<40, True, False)

array([[ True,  True,  True,  True, False,  True],
       [ True,  True, False,  True, False,  True],
       [ True,  True, False, False,  True,  True],
       [ True, False,  True,  True, False,  True],
       [ True,  True,  True, False,  True,  True],
       [ True,  True,  True,  True,  True,  True]])

### np.where -- Case 3
- Syntax : - np.where(condition, x, arr)
- If the condition is True, then x is the result, otherwise the value remains the same

In [62]:
arr2

array([[10, 37, 16,  1, 50, 29],
       [26, 30, 49, 30, 50,  9],
       [10,  1, 43, 41, 37, 17],
       [37, 48, 12, 25, 44, 34],
       [ 9, 37, 15, 50, 14,  6],
       [14, 26, 14, 29, 23, 31]])

In [63]:
np.where(arr2%2, "even", "arr2")

array([['arr2', 'even', 'arr2', 'even', 'arr2', 'even'],
       ['arr2', 'arr2', 'even', 'arr2', 'arr2', 'even'],
       ['arr2', 'even', 'even', 'even', 'even', 'even'],
       ['even', 'arr2', 'arr2', 'even', 'arr2', 'arr2'],
       ['even', 'even', 'even', 'arr2', 'arr2', 'arr2'],
       ['arr2', 'arr2', 'arr2', 'even', 'even', 'even']], dtype='<U4')

In [64]:
np.where(arr2<40,1,arr2)

array([[ 1,  1,  1,  1, 50,  1],
       [ 1,  1, 49,  1, 50,  1],
       [ 1,  1, 43, 41,  1,  1],
       [ 1, 48,  1,  1, 44,  1],
       [ 1,  1,  1, 50,  1,  1],
       [ 1,  1,  1,  1,  1,  1]])

### Replace the Null Values in an array

In [105]:
arr7

array([ 12.,   1.,   1.,  55.,   1.,  99.,   1.,  76.,   1.,  45., 345.,
         1.])

In [106]:
arr7[np.isnan(arr7)] = 1
arr7

array([ 12.,   1.,   1.,  55.,   1.,  99.,   1.,  76.,   1.,  45., 345.,
         1.])

In [107]:
arr7.astype(int)

array([ 12,   1,   1,  55,   1,  99,   1,  76,   1,  45, 345,   1])

### Replacing all the Null values with means , medians
- Both are the ways : arr[np.isnan(arr) == False].mean()
- arr[np.isnan(arr)] = arr(~np.isnan(arr)).mean()
#### By mean 

In [108]:
arr2.mean()

26.5

In [109]:
m = arr2[np.isnan(arr2) == False].mean()
m

26.5

In [110]:
arr2

array([[10, 37, 16,  1, 50, 29],
       [26, 30, 49, 30, 50,  9],
       [10,  1, 43, 41, 37, 17],
       [37, 48, 12, 25, 44, 34],
       [ 9, 37, 15, 50, 14,  6],
       [14, 26, 14, 29, 23, 31]])

In [111]:
a = np.array([ 1 ,2 , 3 , np.nan , 5 ,6 , np.nan , np.nan , 76]).reshape(3,3)
a

array([[ 1.,  2.,  3.],
       [nan,  5.,  6.],
       [nan, nan, 76.]])

In [112]:
a.mean()

nan

In [113]:
y = a[np.isnan(a) == False].mean()
y

15.5

In [114]:
a[np.isnan(a)]=y
a

array([[ 1. ,  2. ,  3. ],
       [15.5,  5. ,  6. ],
       [15.5, 15.5, 76. ]])

### By Median

In [115]:
a

array([[ 1. ,  2. ,  3. ],
       [15.5,  5. ,  6. ],
       [15.5, 15.5, 76. ]])

In [116]:
np.median(a)

6.0

In [117]:
m = a[np.isnan(a)==False]
np.median(m)

6.0

In [118]:
a[np.isnan(a)] = np.median(m)
a

array([[ 1. ,  2. ,  3. ],
       [15.5,  5. ,  6. ],
       [15.5, 15.5, 76. ]])

In [133]:
z = np.array([ 12 , np.nan , 24 , np.nan , 36 , np.nan , 48 , np.nan , 60 , np.nan , 72, 84])
z

array([12., nan, 24., nan, 36., nan, 48., nan, 60., nan, 72., 84.])

In [134]:
np.mean(z)

nan

In [135]:
meannn = z[np.isnan(z) == False].mean()
meannn

48.0

In [136]:
z[np.isnan(z)] = meannn
z

array([12., 48., 24., 48., 36., 48., 48., 48., 60., 48., 72., 84.])

In [137]:
np.median(z)

48.0

In [138]:
v = z[np.isnan(z) == False]
v

array([12., 48., 24., 48., 36., 48., 48., 48., 60., 48., 72., 84.])

In [139]:
z.astype(int)

array([12, 48, 24, 48, 36, 48, 48, 48, 60, 48, 72, 84])

### Replacing Null values By mean

In [140]:
arr1 = np.array([np.nan , 32 , 23 , np.nan , np.nan , np.nan , 76 , 87 , np.nan])
arr1

array([nan, 32., 23., nan, nan, nan, 76., 87., nan])

In [141]:
print(arr1.mean())
print("----------------------------------------------------------")
w = arr1[np.isnan(arr1) == False].mean()
print(w)
print("----------------------------------------------------------")
arr1[np.isnan(arr1)] = w
print(arr1)

nan
----------------------------------------------------------
54.5
----------------------------------------------------------
[54.5 32.  23.  54.5 54.5 54.5 76.  87.  54.5]


### Replacing Null values By Median

In [142]:
arr9 = np.array([np.nan , 101 , 223 , np.nan , np.nan , np.nan , 676 , 787 , np.nan])
arr9

array([ nan, 101., 223.,  nan,  nan,  nan, 676., 787.,  nan])

In [143]:
print(np.median(arr9))
print("-------------------------------------------------------")
qqq = arr9[np.isnan(arr9) == False]
x = np.median(qqq)
print(x)
print("-------------------------------------------------------")
arr9[np.isnan(arr9)] = x
print(arr9)

nan
-------------------------------------------------------
449.5
-------------------------------------------------------
[449.5 101.  223.  449.5 449.5 449.5 676.  787.  449.5]


##### As you can see in below code , I've replaced non null values with 22 and null values with zero(0)

In [130]:
arr9[np.isnan(arr9) == False]= 22
arr9[np.isnan(arr9) == True] = 0
arr9

array([22., 22., 22., 22., 22., 22., 22., 22., 22.])

In [131]:
arr9.astype(int)

array([22, 22, 22, 22, 22, 22, 22, 22, 22])

### np.argwhere()
- It returns an array object that contains the indexes of all the non zero element from the array
##### Syntax:
- np.argwhere(arr)

In [132]:
arr9

array([22., 22., 22., 22., 22., 22., 22., 22., 22.])

In [144]:
np.argwhere(arr9)

array([[0],
       [1],
       [2],
       [3],
       [4],
       [5],
       [6],
       [7],
       [8]], dtype=int64)

In [145]:
arr1

array([54.5, 32. , 23. , 54.5, 54.5, 54.5, 76. , 87. , 54.5])

In [146]:
np.argwhere(arr1)

array([[0],
       [1],
       [2],
       [3],
       [4],
       [5],
       [6],
       [7],
       [8]], dtype=int64)

#### As you can see that in this array there is no zero elements so it giving all the indexes

In [147]:
argg = np.array([ 1, 2, 3, 4, 5, 0, 0, 78, 9, 0, 7, 89, 0, 222, 0])
print(argg)
print("-------------------------------------------------------------------")
print("Non - Zero elements Indexes : " , np.argwhere(argg))
print("-------------------------------------------------------------------")
print("Zero Elements Indexes : ", np.argwhere(argg == 0))

[  1   2   3   4   5   0   0  78   9   0   7  89   0 222   0]
-------------------------------------------------------------------
Non - Zero elements Indexes :  [[ 0]
 [ 1]
 [ 2]
 [ 3]
 [ 4]
 [ 7]
 [ 8]
 [10]
 [11]
 [13]]
-------------------------------------------------------------------
Zero Elements Indexes :  [[ 5]
 [ 6]
 [ 9]
 [12]
 [14]]


### np.argmax()
- It returns the index of the maximum number
- In case of multiple maximum number, it will return the index of the first maximum number
- It treats the n-d array as 1-D array

In [148]:
arr1

array([54.5, 32. , 23. , 54.5, 54.5, 54.5, 76. , 87. , 54.5])

In [149]:
np.argmax(arr1)

7

##### As you can see , it is giving index 7th of the arr1 which is 87 . as you can see it treats the array ino 1D array.

### np.argmin()
- It returns the index of the minimum number
- In case of multiple minimum number, it will return the index of the first minimum number
- It treats the n-d array as 1-D array

In [150]:
arr1

array([54.5, 32. , 23. , 54.5, 54.5, 54.5, 76. , 87. , 54.5])

In [151]:
np.argmin(arr1)

2

#### As you can see , it is giving index 2nd of the arr1 which is 23 . as you can see it treats the array ino 1D array.

### Capping The Value

In [152]:
np.random.seed(10)
arrg=np.random.randint(1,51,36).reshape(6,6)
arrg

array([[10, 37, 16,  1, 50, 29],
       [26, 30, 49, 30, 50,  9],
       [10,  1, 43, 41, 37, 17],
       [37, 48, 12, 25, 44, 34],
       [ 9, 37, 15, 50, 14,  6],
       [14, 26, 14, 29, 23, 31]])

In [153]:
#### cap the value of the above the above array between the range 40 -70 
#### any value less than 40-->40 || any value greater than 70 --->70

In [154]:
np.where(arrg>70,70,np.where(arrg<40,40,arrg))

array([[40, 40, 40, 40, 50, 40],
       [40, 40, 49, 40, 50, 40],
       [40, 40, 43, 41, 40, 40],
       [40, 48, 40, 40, 44, 40],
       [40, 40, 40, 50, 40, 40],
       [40, 40, 40, 40, 40, 40]])

In [155]:
np.random.seed(1)
arrg2 = np.random.randint(100,510,36).reshape(6,6)
arrg2

array([[137, 335, 496, 172, 355, 493],
       [303, 233, 435, 244, 229, 171],
       [337, 490, 381, 278, 376, 354],
       [457, 502, 495, 352, 256, 498],
       [150, 168, 315, 341, 452, 186],
       [241, 493, 107, 419, 417, 122]])

##### cap the value of the above the above array between the range 250 any value less than 250-->"Less Than 250" || any value greater than 250 --->"Greater than 250"

### np.clip()
- It is used to clipping the value between lower limit to upper limit.
#### Syntax
- np.clip(arr , lower_limit , upper_limit)

In [162]:
arrg

array([[10, 37, 16,  1, 50, 29],
       [26, 30, 49, 30, 50,  9],
       [10,  1, 43, 41, 37, 17],
       [37, 48, 12, 25, 44, 34],
       [ 9, 37, 15, 50, 14,  6],
       [14, 26, 14, 29, 23, 31]])

In [163]:
np.clip(arrg,20,30)

array([[20, 30, 20, 20, 30, 29],
       [26, 30, 30, 30, 30, 20],
       [20, 20, 30, 30, 30, 20],
       [30, 30, 20, 25, 30, 30],
       [20, 30, 20, 30, 20, 20],
       [20, 26, 20, 29, 23, 30]])

In [164]:
arr2

array([[10, 37, 16,  1, 50, 29],
       [26, 30, 49, 30, 50,  9],
       [10,  1, 43, 41, 37, 17],
       [37, 48, 12, 25, 44, 34],
       [ 9, 37, 15, 50, 14,  6],
       [14, 26, 14, 29, 23, 31]])

In [165]:
np.clip(arrg2,30,50)

array([[50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50],
       [50, 50, 50, 50, 50, 50]])

In [167]:
np.random.seed(7)
arr = np.random.randint(1,51,36).reshape(6,6)
arr

array([[48,  5, 26,  4, 20, 24],
       [40, 29, 15, 24,  9, 26],
       [47, 43, 27,  9, 40, 39],
       [ 5, 49,  8, 45,  1, 12],
       [ 7, 20, 45,  6, 25, 49],
       [ 1, 35, 10, 15, 39,  5]])

In [168]:
np.diag(arr)

array([48, 29, 27, 45, 25,  5])

In [169]:
np.diag(np.diag(arr))

array([[48,  0,  0,  0,  0,  0],
       [ 0, 29,  0,  0,  0,  0],
       [ 0,  0, 27,  0,  0,  0],
       [ 0,  0,  0, 45,  0,  0],
       [ 0,  0,  0,  0, 25,  0],
       [ 0,  0,  0,  0,  0,  5]])

In [170]:
arr

array([[48,  5, 26,  4, 20, 24],
       [40, 29, 15, 24,  9, 26],
       [47, 43, 27,  9, 40, 39],
       [ 5, 49,  8, 45,  1, 12],
       [ 7, 20, 45,  6, 25, 49],
       [ 1, 35, 10, 15, 39,  5]])

In [171]:
print(np.mean(arr))
print(np.var(arr))
print(np.std(arr))

23.666666666666668
253.16666666666666
15.911211979816832


In [172]:
import math
math.sqrt(253.16666666666666)

15.911211979816832

In [173]:
np.median(arr)

24.0

In [175]:
arr3 = np.array([12, np.nan , 31 , np.nan , np.nan , np.nan , 34 ,35 , 89]).reshape(3,3)
arr3

array([[12., nan, 31.],
       [nan, nan, nan],
       [34., 35., 89.]])

In [176]:
arr3[np.isnan(arr3) == False]

array([12., 31., 34., 35., 89.])

In [177]:
meann = arr3[np.isnan(arr3) == False].mean()
meann

40.2

In [178]:
arr3[np.isnan(arr3)] = meann
print(arr3)

[[12.  40.2 31. ]
 [40.2 40.2 40.2]
 [34.  35.  89. ]]


In [180]:
arr3.astype(int)

array([[12, 40, 31],
       [40, 40, 40],
       [34, 35, 89]])

In [181]:
arr

array([[48,  5, 26,  4, 20, 24],
       [40, 29, 15, 24,  9, 26],
       [47, 43, 27,  9, 40, 39],
       [ 5, 49,  8, 45,  1, 12],
       [ 7, 20, 45,  6, 25, 49],
       [ 1, 35, 10, 15, 39,  5]])

In [182]:
arr[1,(0,1,2,3,4,5)] = 22,21,19,15,51,21
arr

array([[48,  5, 26,  4, 20, 24],
       [22, 21, 19, 15, 51, 21],
       [47, 43, 27,  9, 40, 39],
       [ 5, 49,  8, 45,  1, 12],
       [ 7, 20, 45,  6, 25, 49],
       [ 1, 35, 10, 15, 39,  5]])

In [183]:
arr[:: , :1:]

array([[48],
       [22],
       [47],
       [ 5],
       [ 7],
       [ 1]])

### Condition Statements on Array
- It returns boolean values of array. Syntax
    - arr_name[condition]

In [11]:
np.random.seed(12)
arr1 = np.random.randint(22,87,36).reshape(6,6)
arr1

array([[49, 28, 24, 25, 70, 44],
       [71, 74, 27, 35, 56, 22],
       [35, 57, 84, 57, 55, 52],
       [85, 40, 72, 28, 67, 52],
       [54, 49, 81, 81, 47, 78],
       [26, 66, 59, 75, 75, 69]])

##### As you can see in below code that it is giving bool values of arrays because we directly put arr1>25 if we put the same condition inside arr_name then it'll only show the greater elements of arr1 by 25 you can see in below codes.

In [12]:
arr1>25

array([[ True,  True, False, False,  True,  True],
       [ True,  True,  True,  True,  True, False],
       [ True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True]])

In [13]:
arr1[arr1>25]


array([49, 28, 70, 44, 71, 74, 27, 35, 56, 35, 57, 84, 57, 55, 52, 85, 40,
       72, 28, 67, 52, 54, 49, 81, 81, 47, 78, 26, 66, 59, 75, 75, 69])

In [14]:
arr1[arr1<36]

array([28, 24, 25, 27, 35, 22, 35, 28, 26])

### Filtering elements by giving conditions

In [15]:
arr1

array([[49, 28, 24, 25, 70, 44],
       [71, 74, 27, 35, 56, 22],
       [35, 57, 84, 57, 55, 52],
       [85, 40, 72, 28, 67, 52],
       [54, 49, 81, 81, 47, 78],
       [26, 66, 59, 75, 75, 69]])

In [16]:
arr1[(arr1 % 3 ==0) & (arr1 % 2 == 0) ]

array([24, 84, 72, 54, 78, 66])

###  Array Concatination
- Vertical Concatination
- Horizontal Concatination
#### Syntax
- np.concatenate((x,y) , axis = 0,1)

In [1]:
import numpy as np

In [2]:
a = np.arange(24).reshape(4,6)
a

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23]])

In [3]:
a.shape

(4, 6)

In [5]:
b=np.arange(6).reshape(1,6)
b

array([[0, 1, 2, 3, 4, 5]])

In [6]:
np.concatenate((a,b), axis=0)

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23],
       [ 0,  1,  2,  3,  4,  5]])

##### Watch the shape and size of "a" array which is shape(4,6) means 4 rows and 6 columns thats why now we have to make the another array in that suitable way

In [7]:
c = np.arange(4).reshape(4,1)
print(c)

[[0]
 [1]
 [2]
 [3]]


In [8]:
print("------------------------------------------")
print(np.concatenate((a,c) , axis = 1))

------------------------------------------
[[ 0  1  2  3  4  5  0]
 [ 6  7  8  9 10 11  1]
 [12 13 14 15 16 17  2]
 [18 19 20 21 22 23  3]]


### Arthemetic Operations on Array

In [17]:
arr1

array([[49, 28, 24, 25, 70, 44],
       [71, 74, 27, 35, 56, 22],
       [35, 57, 84, 57, 55, 52],
       [85, 40, 72, 28, 67, 52],
       [54, 49, 81, 81, 47, 78],
       [26, 66, 59, 75, 75, 69]])

In [18]:
arr1+5

array([[54, 33, 29, 30, 75, 49],
       [76, 79, 32, 40, 61, 27],
       [40, 62, 89, 62, 60, 57],
       [90, 45, 77, 33, 72, 57],
       [59, 54, 86, 86, 52, 83],
       [31, 71, 64, 80, 80, 74]])

In [20]:
arr1//3

array([[16,  9,  8,  8, 23, 14],
       [23, 24,  9, 11, 18,  7],
       [11, 19, 28, 19, 18, 17],
       [28, 13, 24,  9, 22, 17],
       [18, 16, 27, 27, 15, 26],
       [ 8, 22, 19, 25, 25, 23]], dtype=int32)

In [21]:
arr1*5

array([[245, 140, 120, 125, 350, 220],
       [355, 370, 135, 175, 280, 110],
       [175, 285, 420, 285, 275, 260],
       [425, 200, 360, 140, 335, 260],
       [270, 245, 405, 405, 235, 390],
       [130, 330, 295, 375, 375, 345]])

In [22]:
arr1%4

array([[1, 0, 0, 1, 2, 0],
       [3, 2, 3, 3, 0, 2],
       [3, 1, 0, 1, 3, 0],
       [1, 0, 0, 0, 3, 0],
       [2, 1, 1, 1, 3, 2],
       [2, 2, 3, 3, 3, 1]], dtype=int32)

In [23]:
arr1-4

array([[45, 24, 20, 21, 66, 40],
       [67, 70, 23, 31, 52, 18],
       [31, 53, 80, 53, 51, 48],
       [81, 36, 68, 24, 63, 48],
       [50, 45, 77, 77, 43, 74],
       [22, 62, 55, 71, 71, 65]])

###### As you can see all these arthemetic operations are done over arrays which is basically asked by the interviewer how these operations are handled by numpy.
###### That where we can say that we are using array Broadcasting. 

### Array Broadcasting
- When we perform some array to array operation on an array of different size then to make the operation possible Numpy adds some dumpy rows and dummy columns to match the shape of both the arrays 
- This process is called as Array Broadcasting
- NOTE : - Atleast one dimension should be same and the other dimension hould be 1 in either of the array 

In [24]:
arr1+5

array([[54, 33, 29, 30, 75, 49],
       [76, 79, 32, 40, 61, 27],
       [40, 62, 89, 62, 60, 57],
       [90, 45, 77, 33, 72, 57],
       [59, 54, 86, 86, 52, 83],
       [31, 71, 64, 80, 80, 74]])

### Iterate over an Array

In [25]:
w = np.arange(24)
w

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [26]:
for i in w:
    print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


In [27]:
arr1

array([[49, 28, 24, 25, 70, 44],
       [71, 74, 27, 35, 56, 22],
       [35, 57, 84, 57, 55, 52],
       [85, 40, 72, 28, 67, 52],
       [54, 49, 81, 81, 47, 78],
       [26, 66, 59, 75, 75, 69]])

In [28]:
for i in arr1:
    print(i)

[49 28 24 25 70 44]
[71 74 27 35 56 22]
[35 57 84 57 55 52]
[85 40 72 28 67 52]
[54 49 81 81 47 78]
[26 66 59 75 75 69]


#####  As you can see it is not giving elements over an iteration so we have a function for that in numpy which is known as np.nditer()

### np.nditer()
- It returns an iterable object. 

In [29]:
arr1

array([[49, 28, 24, 25, 70, 44],
       [71, 74, 27, 35, 56, 22],
       [35, 57, 84, 57, 55, 52],
       [85, 40, 72, 28, 67, 52],
       [54, 49, 81, 81, 47, 78],
       [26, 66, 59, 75, 75, 69]])

In [31]:
for i in np.nditer(arr1):
    print(i,end=" ")

49 28 24 25 70 44 71 74 27 35 56 22 35 57 84 57 55 52 85 40 72 28 67 52 54 49 81 81 47 78 26 66 59 75 75 69 

###  Determinant
- It is used to find the determinant of the array.
- Formula
    - det=a(ei−fh)−b(di−fg)+c(dh−eg)
    - Which differs to array like this:
    - [[a,b,c], [d,e,f],[g,h,i]
##### Syntax
- np.linalg.det(arr)]]

In [32]:
arr1

array([[49, 28, 24, 25, 70, 44],
       [71, 74, 27, 35, 56, 22],
       [35, 57, 84, 57, 55, 52],
       [85, 40, 72, 28, 67, 52],
       [54, 49, 81, 81, 47, 78],
       [26, 66, 59, 75, 75, 69]])

In [33]:
np.linalg.det(arr1)

2266430643.0000033

In [34]:
arr2 = np.arange(9).reshape(3,3)
arr2

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [35]:
np.linalg.det(arr2)

0.0